### ISSUE TITLE
Memory leaks when pm2 updated to 3.4.0 or later version

### ISSUE THREAD

**Issue by liujiede** [Unitech/pm2#4302]
## Memory leaks in 3.4.0 and later versions

each request will be an instance of new [Meter.js](https://github.com/keymetrics/pm2-io-apm/blob/master/src/utils/metrics/meter.ts#L30)

### Call chain

* node_modules/@pm2/io/build/main/index.js
```javascript
const pmx_1 = require("./pmx");
const IO_KEY = Symbol.for('@pm2/io');
const isAlreadyHere = (Object.getOwnPropertySymbols(global).indexOf(IO_KEY) > -1);
```

* node_modules/@pm2/io/build/main/pmx.js
```javascript
        this.actionService.init();
        serviceManager_1.ServiceManager.set('actions', this.actionService);
        this.metricService = new metrics_1.MetricService();
```

* node_modules/@pm2/io/build/main/metrics/httpMetrics.js
```javascript
    hookRequire() {
        this.hooks = requireMiddle(['http', 'https'], (exports, name) => {
            this.hookHttp(exports, name);
            return exports;
        });
    }
```
```javascript
    if (event !== 'request')
                    return original.apply(this, arguments);
                if (name === 'http') {
                    self.registerHttpMetric();
                }
```
* node_modules/@pm2/io/build/main/utils/metrics/meter.js
```javascript
    this._interval = setInterval(function () {
      self._rate.tick();
    }, this._tickInterval);
    this._interval.unref();
```
### Simple test case
```javascript
//test.js
const http = require('http');
http.createServer(function(req, res){
  res.end('okay');
}).listen(8080)
```
1、run by `./node_modules/.bin/pm2-dev  start ./test.js`

2、request `http://127.0.0.1:8080`

**as the amount of requests increases, the cpu will eventually skyrocket and memory leaks**

**Comment by ZhangDianPeng** [comment]
I also encountered this problem when updated to 3.5.1.

**Comment by rchl** [comment]
Is that with `@pm2/io` >= 4.2.2?

**Comment by liujiede** [comment]
> Is that with `@pm2/io` >= 4.2.2?

--------------------------
Seems to be >=4.2.1
 
>no problem
```javascript
private hookRequire () {
    //...
    shimmer.wrap(Module, '_load', function (original: Function) {
      return function (file) {
        // we first require the target module
        const returned = original.apply(original, arguments)
        // if it's a modue that we want to patch, install the patch on top of it
        if (file === 'http' && self.modules['http'] === undefined) {
          self.registerHttpMetric()
          self.hookHttp(returned, file)
        } else if (file === 'https' && self.modules['https'] === undefined) {
          self.registerHttpsMetric()
          self.hookHttp(returned, file)
        }
        // and of course send the module so user can use it
        return returned
      }
    })
    //...
  }
```

>problematic
```javascript
 private hookRequire () {
    this.hooks = requireMiddle(['http', 'https'], (exports, name) => {
      this.hookHttp(exports, name)
      return exports
    })
  }
```

**Comment by Eywek** [comment]
You can reproduce with 4.2.2?

**Comment by pengpwang** [comment]
I also encountered this problem when updated to pm2@3.5.1.

**Comment by hedgeday** [comment]
I encountered the problem when I updated to 3.5.1, but can't seem to reproduce with 3.5.0.

**Comment by rchl** [comment]
Version of pm2 doesn't really matter as the problem is/was in @pm2/io package and installing any recent version of pm2 always installs latest version of@pm2/io.
The bug might be fixed already in latest version of@pm2/io btw. I'm still verifying that.

**Comment by HughePaul** [comment]
Is there a way to disable the HTTP and HTTPS intercepts completely? This is causing memory and CPU runaway in production at the moment

**Comment by rchl** [comment]
Memory leak is fixed as long as your pm2 is using `@pm2/io` >= 4.2.2

**Comment by wizardpisces** [comment]
Also this problem when updated to pm2@3.5.1, downgrade to pm2@3.2.2 fixed the leak for now

**Comment by hodoublesy** [comment]
I am still seeing CPU and Memory usage stair-step up even on 4.2.3 - going to try downgrading pm2@3.2.2 to see if it fixes this.

**Comment by maxfadeev** [comment]
@hodoublesy, did that help you ?

**Comment by hodoublesy** [comment]
@maxfadeev - yes, after downgrading to 3.2.2 everything has been working great.  I've had runaway memory leak issues in the past upgrading to the latest pm2 so I'll just be cautious in the future upgrading!

**Comment by tuananh** [comment]
is there any update to fix this?

**Comment by Unitech** [comment]
This has been fixed with @pm2/io 4.2.2
